source/val/ValidationState.h

```diff
@@ -138,36 +144,58 @@ class ValidationState_t {
   /// Keeps track of ID definitions and uses.
   class UseDefTracker {
    public:
-    void AddDef(const spv_id_info_t& def) { defs_[def.id] = def; }
+    explicit UseDefTracker(ValidationState_t* state);
+
+    /// Adds the defintion of an ID
+    void AddDef(const spv_id_info_t& def);
 
-    void AddUse(uint32_t id) { uses_.insert(id); }
+    /// Adds the use of an ID
+    /// @param[in] id      The id of the operand
+    /// @param[in] is_phi  True if the use of the id was part of a phi
+    ///                    instruction
+    void AddUse(uint32_t id, bool is_phi);
 
     /// Finds id's def, if it exists.  If found, returns <true, def>. Otherwise,
     /// returns <false, something>.
-    std::pair<bool, spv_id_info_t> FindDef(uint32_t id) const {
-      if (defs_.count(id) == 0) {
-        return std::make_pair(false, spv_id_info_t{});
-      } else {
-        /// We are in a const function, so we cannot use defs.operator[]().
-        /// Luckily we know the key exists, so defs_.at() won't throw an
-        /// exception.
-        return std::make_pair(true, defs_.at(id));
-      }
-    }
+    std::pair<bool, spv_id_info_t> FindDef(uint32_t id) const;
+
+    /// Propagates the definitions of IDs from the dominators to the blocks they
```

Comment:
The blurb is a little confusing.  I'd say

"Updates bookkeeping to record that an ID definition in a given block is visible in blocks it dominates."

The note should also say that it should only be run once, otherwise you'll get redundant data.
?